packages/wonder-blocks-form/src/components/text-area.tsx

```diff
@@ -139,6 +139,10 @@ type TextAreaProps = AriaProps & {
      * Called right after the textarea is validated.
      */
     onValidate?: (errorMessage?: string | null | undefined) => unknown;
+    /**
+     * Whether the textarea is in an error state.
+     */
+    error?: boolean;
```

Comment:
I feel that we could get into a tricky situation here.

Have you found a situation where a component instance needs to use both `error` and `validate`?

I'm thinking if we need to add more guardrails to prevent a case where these two props could collide.

It's tricky from the API perspective that the validation function would pass but the user would still see an error, and I wonder if this case is expected or not (and if not, then we should prevent from this to happen from the dev perspective).

Here's an example:
https://5e1bf4b385e3fb0020b7073c-kwxacxtfdq.chromatic.com/iframe.html?args=error:!true&id=packages-form-textarea--error-from-validation&viewMode=story

<img width="1274" alt="Screenshot 2024-10-16 at 11 23 04 AM" src="https://github.com/user-attachments/assets/38ac04a4-861d-4fe3-9b0e-feee62e122dd">

The email pattern is now valid, but the input is still showing its `error` state.

I wonder if/how we need to cover this case. wdyt?

Reply:
Great question! I think this scenario is behaving as I would have expected (remaining in an error state after the email is in a valid format because the error prop is set to true). An example use case for this could be a sign up form where even if the email is valid, there could still be an error after form submission like the email provided is already associated with an account. 

The workaround I used for backend validation errors before was to keep track of whether there is an error in the component state, check for that error in the validate function, and then also set the `key` on the component to whether there was an error so that the component would re-mount to trigger the validate function. Here is an [example from webapp](https://github.com/Khan/webapp/pull/24613/files#r1717640619) where I ran into this issue! 

Part of the confusion could also be there is no error message in the story. The error message isn't part of the component, but it is shown in the story for example purposes. I'll update the story so another error message is shown if the `error` control is set! 

Let me know what you think or if you have some ideas on other alternatives!

Reply:
oh I see, thanks for the context.

Given your example, I still see that the user could get into an state where they use an email that is not associated with an account and they use valid email. Should the input update to its valid state? or should it remain using the error state?

Reply:
The input would return to a valid state when the `error` prop is set to false (since in that case the email is fine to use).

I'll put together another story that has both `error` and `validate` props configured with logic so we can see how it will behave!

Reply:
> The input would return to a valid state when the error prop is set to false (since in that case the email is fine to use).

That's the point I'm trying to get into.... this could lead into an "[make impossible states impossible](https://kentcdodds.com/blog/make-impossible-states-impossible)" situation. We would need to make sure that both `error` and `validate` are in sync and there might be cases where it would be confusing for the developer implementing a feature with these characteristics.

I wonder if there's a way to refactor this API to make it more predictive (even if it means changing `validate/onValidate`).

Reply:
Thanks for the feedback, Juan! From our pairing session, we decided that:
- it makes sense to support both `error` and `validate` props in the TextField and TextArea components since there are use cases for using both
- we will need documentation around when to use these props
- we will leave validation up to the field components instead of having LabeledField pass it down. This means LabeledField will only have an error prop for the error message and the validate prop will stay on TextField/TextArea
- it will be helpful to have a shared utility/hook for the shared logic around validation: when to trigger validation, support for async validation, etc (will be explored in another PR). 

I've added some more documentation around when to use `error` vs `validate` props, and have added examples for when both are used ([TextField example](https://5e1bf4b385e3fb0020b7073c-cvpbgtsyrn.chromatic.com/?path=/story/packages-form-textfield--error-from-prop-and-validation) and [TextArea example](https://5e1bf4b385e3fb0020b7073c-cvpbgtsyrn.chromatic.com/?path=/story/packages-form-textarea--error-from-prop-and-validation))! I'll update the implementation spec as well!

Let me know what you think!